[Pathfinding] Fixed slow ClosedSet contains queries.

tickinginstant opened the pull request:
I just realized that the pathfinding code is currently storing the Closed Set as a list. The only operation it's ever performing on that list is Contains, which is very slow (asymptotically O(n)) on CSharp's built-in list. This changes it to a dictionary, which can check containment much faster (close to O(1)).

I'm pretty sure this drastically speeds up pathfinding. I'll try to get my pathfinding tests working again so that I can actually give some hard numbers, but in the one test that I've managed to get working, it changed one of the searches from ~2.5s to ~0.25s.

EDIT: Thanks to a suggestion by @Mizipzor, I changed from a Dictionary to a HashSet. It's about the same speed, but it's more elegant. Also, I got the tests working again, with my results posted in the comments below. Eight of the twelve tests I used are over 10x faster, and the two slower ones seem to fall within standard deviation.

Mizipzor commented:
If it really is a set (it should be), then why not make it a `HashSet`? Thats even faster, isnt it?

tickinginstant commented:
@Mizipzor I didn't realize that Unity had HashSets! I'll make the change.

tickinginstant commented:
@Mizipzor Unfortunately, some of the more recent changes seem to conflict with my pathfinding tests. I'm still trying to figure out exactly what changed, and why it's conflicting. (Some of it is obvious, like the World script in the Models folder trying to access the WorldController. Some of it I'm having more trouble with.)

I'll try to get some hard numbers as soon as I can, since I think this is a pretty important fix to make.

tickinginstant commented:
All right, I did some quick-and-dirty hacks to get the test code working. (The version on GitHub is still broken, though, since the changes that I made aren't even remotely committable. Think "commenting out swaths of code.") The results:

test name|master|with patch|% cost reduction
----|-----|------|-------
[Empty] Center to Nearby|0.001673937|0.001830268|-9.34%
[Empty] Center to Corner|0.001184654|0.002834868|-139.3%
[Empty] Center to Nearby Inventory|0.01149449|0.003152609|72.57%
[Empty] Center to Corner Inventory|8.637645|0.243082|97.19%
[Switchback] Center to Nearby|0.08960342|0.006751156|92.47%
[Switchback] Center to Corner|2.253765|0.03754053|98.33%
[Switchback] Center to Nearby Inventory|0.1576324|0.01094742|93.06%
[Switchback] Center to Corner Inventory|2.270219|0.03168125|98.6%
[Diagonals] Center to Nearby|0.06740265|0.009619045|85.73%
[Diagonals] Center to Corner|0.5767395|0.0457469|92.07%
[Diagonals] Center to Nearby Inventory|1.163605|0.04426956|96.2%
[Diagonals] Center to Corner Inventory|10.56312|0.1614793|98.47%

I'm not sure why it's so much slower on that second test, but it's a LOT faster on the rest.

Mizipzor commented:
Oh man I fucking love tables. What does the labels mean? Empty, switchback and diagonals?

NogginBops commented:
@tickinginstant [Empty] Center to Corner thats weird...

NogginBops commented:
This is the switchback pathfinding test.
![](https://cloud.githubusercontent.com/assets/20977630/17834009/0fbb6e16-66fd-11e6-9db1-f76f5439023f.png)

Mizipzor commented:
Those were the test with the `Dictionary`, right? Would it be possible to add a column for when a `HashSet` is used?

NogginBops commented:
@Mizipzor The "master" column is the `Dictionary` and the "with patch" is the `HashSet`

Mizipzor commented:
Alright, Im just a bit slow thats all. Good that the code isnt, though. ^^

tickinginstant commented:
@Mizipzor Empty means that the pathfinding test is taking place in a completely blank world, with no walls. Switchback means that it's taking place in a world with a big box around it, and a single 4-wide corridor winding back and forth within the box.

Diagonals means it's placed in a world with some diagonal obstacles -- in particular, two long parallel diagonal lines surrounding the center, going from the lower left corner to the upper right corner, with an extra diagonal wall separating the center from the lower-left corner. (It kind of looks like a cup rotated by 45 degrees.) It's a new test that I added, which basically forces the pathfinding to route through the upper-right corner.

@Mizipzor @NogginBops The "master" column is the version with the original list. The "with patch" column is with HashSet. (I never tested Dictionary, because I changed it to HashSet before I got the tests working again. Should I test Dictionary?)

Also, sorry I didn't explain my tables better!

@NogginBops I think it might be just standard deviation? Those tests that I ran were 10 trials each, but I just ran the tests 100 times for the new code (which I can actually do, because it doesn't take 10 seconds per test! :D), and got these results:

test name | average time | standard deviation
-----|----|----
[Empty] Center to Nearby | 0.0003621912s | 0.001252208
[Empty] Center to Corner | 0.001205401s | 0.002082388
[Empty] Center to Nearby Inventory | 0.003318415s | 0.001739884
[Empty] Center to Corner Inventory | 0.2412141s | 0.01044899
[Switchback] Center to Nearby | 0.007830391s | 0.004188197
[Switchback] Center to Corner | 0.03813854s | 0.007403263
[Switchback] Center to Nearby Inventory | 0.008980675s | 0.004792834
[Switchback] Center to Corner Inventory | 0.03141167s | 0.007416994
[Diagonals] Center to Nearby | 0.007291222s | 0.003591469
[Diagonals] Center to Corner | 0.04851166s | 0.006581025
[Diagonals] Center to Nearby Inventory | 0.04413746s | 0.006424839
[Diagonals] Center to Corner Inventory | 0.1562434s | 0.01050877

So the first two tests are more in line with the pre-patch numbers.

Mizipzor commented:
Im curios what Dictionary gets, and it should be a trivial search and replace in that file, right?

tickinginstant commented:
@Mizipzor Okay, I ran 100 trials for Dictionary and compared them with the results for 100 trials of HashSet. It doesn't really look like there's a significant difference:

Test Name | HashSet | Dictionary | % Improvement with Dictionary
-----|-----|-----|-----
[Empty] Center to Nearby | 0.0003621912 | 0.0003947163 | -8.98%
[Empty] Center to Corner | 0.001205401 | 0.001195068 | 0.86%
[Empty] Center to Nearby Inventory | 0.003318415 | 0.003443799 | -3.78%
[Empty] Center to Corner Inventory | 0.2412141 | 0.2420699 | -0.35%
[Switchback] Center to Nearby | 0.007830391 | 0.007722244 | 1.38%
[Switchback] Center to Corner | 0.03813854 | 0.03745857 | 1.78%
[Switchback] Center to Nearby Inventory | 0.008980675 | 0.00881691 | 1.82%
[Switchback] Center to Corner Inventory | 0.03141167 | 0.03083641 | 1.83%
[Diagonals] Center to Nearby | 0.007291222 | 0.00696724 | 4.44%
[Diagonals] Center to Corner | 0.04851166 | 0.04752083 | 2.04%
[Diagonals] Center to Nearby Inventory | 0.04413746 | 0.0431041 | 2.34%
[Diagonals] Center to Corner Inventory | 0.1562434 | 0.1505671 | 3.63%

All of those differences fall squarely within standard deviation, so I'm inclined to think they're roughly equivalent. But HashSet is probably more correct from a conceptual standpoint, so I'm sticking with it. (Thanks for the correction! :))

Mizipzor commented on Assets/Scripts/Pathfinding/Path_AStar.cs:
```diff
@@ -121,7 +121,7 @@ public Path_AStar(World world, Tile tileStart, Tile tileEnd, string objectType =
                 Path_Node<Tile> neighbor = edge_neighbor.node;
 
-                if (ClosedSet.Contains(neighbor) == true)
+                if (ClosedSet.Contains(neighbor))
```
Thank you for this by the way, checks like that makes my eyes itch.

Sommerbo replied:
Quill18 wrote it like that to make it easier to read. But I agree ==true on a bool is itch inducing.

Mizipzor commented:
Awesome. :) And youre right, an actual `Set` is absolutely right here. What was the bool used for anyway? No matter, I could check the code, the important bit is that this is a massive improvement that needs merging right now.

vogonistic commented:
Merged!

tickinginstant commented:
Awesome! :)
